### ISSUE TITLE
Better destructuring support

### ISSUE THREAD

**Issue by stephencweiss** [stephencweiss/gatsby-source-xkcd#2]
Make it easier to know what kind of data a user can expect when they user query xkcd - typehinting? typescript?

**Comment by Js-Brecht** [comment]
Hey, was just checking out a couple things, and saw this issue.  I hope you don't mind me putting in my two cents.

Mostly, this kind of thing can be handled by the plugin consumer, by using plugins like [`gatsby-plugin-typegen`](https://www.gatsbyjs.org/packages/gatsby-plugin-typegen/) to generate types for their Gatsby GraphQL schema.  Using that plugin in conjunction with [`eslint`](https://www.gatsbyjs.org/packages/gatsby-plugin-typegen/#eslint), [`ts-graphql-plugin`](https://www.gatsbyjs.org/packages/gatsby-plugin-typegen/#typescript-plugin), and [`vscode-apollo`](https://www.gatsbyjs.org/packages/gatsby-plugin-typegen/#vscode-extension) is a very powerful combination.

One thing you could do is make it easier for the user to generate GraphQL queries by creating fragments, like you see in [`gatsby-transformer-sharp`](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sharp/src/fragments.js).

You can also explicitly define your GraphQL types in [`createSchemaCustomization`](https://www.gatsbyjs.org/docs/schema-customization/), which makes it easy for people to go look at your repository, and see what kind of data they might expect to receive from the plugin.  You can also include the types you write there in your `README.md`, just as they are, which is always handy when somebody first starts checking out the plugin.

Other than that, you can ship with types which describe the shape of the data, that the user can then import and use in their project.  Seems strange to me to do 
```ts
import { IFooType } from 'gatsby-source-xkcd';
```
though, but that kind of thing is not unheard of.  Some plugins ship with utilities that the user can use in their source, so their usage in that respect is not isolated to `gatsby-config`.

However, since this is GraphQL we're talking about, the user can query as much, or as little, data as they want.  So, shipping your own types does a couple of things...
* it either means the user is getting types for data that they're _possibly_ not even going to have returned to them, based on their GraphQL query... which makes room for bugs, because they could then wind up trying to use data that isn't there, since they're relying on the contract Typescript provides regarding the shape of the data.
* it makes them transform the imported types themselves, to match the shape of the GraphQL.  Which becomes a maintenance nightmare: if you ever want to change your GraphQL queries, you also have to remember to go change your types.

If you're interested in shipping types for the data returned from your plugin, you can write GraphQL types in a different file (which you can import and use in `createSchemaCustomization`), then use something like [`graphql-code-generator`](https://github.com/dotansimha/graphql-code-generator#readme) (`@graphql-codegen/typescript`, to be precise) during your build to generate the typescript interfaces automatically.

**Comment by stephencweiss** [comment]
Thanks @Js-Brecht! Amazing. Don't mind the $.02 at all!

I think between creating this issue and publishing 1.x, I added types. The most important of which is the [`comic`](https://github.com/stephencweiss/gatsby-source-xkcd/blob/master/src/fetchNodes.ts#L12)

```javascript
export type Comic = {
    num: number;
    alt?: string;
    day?: string;
    img?: string;
    link?: string;
    month?: string;
    news?: string;
    safe_title?: string;
    title?: string;
    transcript?: string;
    year?: string;
};
```

Is the idea with `createSchemaCustomization` that that would be more explicit and in a location that would be more appropriate for people to look? 

The shape of the data is so simple that I'm not sure fragments make much sense (unless I'm misunderstanding them - very possible!). 

The more and more comfortable I get with GraphiQL, the less necessary it seems to be - exploring the data is just so easy. I'm just trying to be considerate of folks who aren't great with exploring a GraphQL layer (like me! ha).

Thanks again for weighing in - learning a ton from you and really appreciate it!

**Comment by Js-Brecht** [comment]
> Is the idea with `createSchemaCustomization` that that would be more explicit and in a location that would be more appropriate for people to look?

Yeah, pretty much.  Gatsby is perfectly able to infer the types from the shape of the data.  Writing your own statically defined types is essentially just self-documenting.  Having an explicit Typescript interface does the same thing.

I just mention that method because then you could have Typescript interfaces generated automatically, by the GraphQL types that are drawn up, which just reduces the amount of work you'd have to do for maintenance.  But really, this is probably only applicable if your data set was in flux, or if you had to leverage `createSchemaCustomization` to get some particular benefits (like directives, custom types, nullable/non-nullable fields, etc...).  If you find yourself needing to use it, then it would make sense to only write the types in one place, and let the rest happen automatically.

You could probably have a unit test or something that could run, which could query data, and then generate the types based off the inferred schema, as well; that would have an added benefit of ensuring your code works as expected.  But that might be more work than it's worth, if generating the schema is the only purpose.

> The shape of the data is so simple that I'm not sure fragments make much sense (unless I'm misunderstanding them - very possible!).

Yeah, probably not.  It would make more sense if you had, potentially, specific data structures that should be used for definite purposes in various different contexts.  Saves time and hassle having to write out fields over and over.
